File: church/church/report/visitor_conversion/visitor_conversion.js
// Copyright (c) 2026, kunle and contributors
// For license information, please see license.txt
/* eslint-disable */

// church/church/report/visitor_conversion/visitor_conversion.js

frappe.query_reports["Visitor Conversion"] = {

    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_months(frappe.datetime.now_date(), -12)
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.now_date()
        },
        {
            fieldname: "period",
            label: __("Group By"),
            fieldtype: "Select",
            options: "Monthly\nQuarterly\nYearly",
            default: "Monthly"
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch"
        }
    ],

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (!data) return value;

        // Period — bold blue
        if (column.fieldname === "period") {
            value = `<strong style="color:#2E75B6;">${data.period || ""}</strong>`;
        }

        // Total visitors — bold
        if (column.fieldname === "total_visitors") {
            value = `<strong>${data.total_visitors || 0}</strong>`;
        }

        // Converted — green
        if (column.fieldname === "converted") {
            const n = parseInt(data.converted) || 0;
            value = n > 0
                ? `<span style="color:#27ae60;font-weight:700;">✅ ${n}</span>`
                : `<span style="color:#95a5a6;">0</span>`;
        }

        // Lost contact — red
        if (column.fieldname === "lost_contact") {
            const n = parseInt(data.lost_contact) || 0;
            value = n > 0
                ? `<span style="color:#e74c3c;">❌ ${n}</span>`
                : `<span style="color:#95a5a6;">0</span>`;
        }

        // In follow-up — orange
        if (column.fieldname === "in_followup") {
            const n = parseInt(data.in_followup) || 0;
            value = n > 0
                ? `<span style="color:#e67e22;font-weight:600;">🔄 ${n}</span>`
                : `<span style="color:#95a5a6;">0</span>`;
        }

        // Conversion rate — progress bar + %
        if (column.fieldname === "conversion_rate") {
            const pct = parseFloat(data.conversion_rate) || 0;
            const col = pct >= 20 ? "#27ae60" : pct >= 10 ? "#f39c12" : "#e74c3c";
            value = `<div style="display:flex;align-items:center;gap:6px;">
                         <div style="flex:1;background:#ecf0f1;border-radius:4px;
                                     height:10px;min-width:60px;">
                             <div style="width:${Math.min(pct * 3, 100)}%;background:${col};
                                         height:10px;border-radius:4px;"></div>
                         </div>
                         <span style="color:${col};font-weight:700;font-size:12px;
                                      min-width:40px;">${pct}%</span>
                     </div>`;
        }

        return value;
    },

    onload: function(report) {
        report.page.add_inner_button(__("📅 Last 12 Months"), function() {
            frappe.query_report.set_filter_value("from_date", frappe.datetime.add_months(frappe.datetime.now_date(), -12));
            frappe.query_report.set_filter_value("to_date",   frappe.datetime.now_date());
            frappe.query_report.set_filter_value("period",    "Monthly");
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("📅 This Year"), function() {
            frappe.query_report.set_filter_value("from_date", new Date().getFullYear() + "-01-01");
            frappe.query_report.set_filter_value("to_date",   frappe.datetime.now_date());
            frappe.query_report.set_filter_value("period",    "Quarterly");
            frappe.query_report.refresh();
        });
    }
};

File: church/church/report/visitor_conversion/visitor_conversion.py
# Copyright (c) 2026, kunle and contributors
# For license information, please see license.txt

# church/church/report/visitor_conversion/visitor_conversion.py

import frappe
from frappe import _
from frappe.utils import nowdate, add_to_date

def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data    = get_data(filters)
    summary = get_summary(data)
    chart   = get_chart(data)
    return columns, data, None, chart, summary

def get_columns():
    return [
        {"fieldname": "period",           "label": _("Period"),             "fieldtype": "Data", "width": 120},
        {"fieldname": "total_visitors",   "label": _("Total Visitors"),     "fieldtype": "Int",  "width": 120},
        {"fieldname": "first_timers",     "label": _("First Timers"),       "fieldtype": "Int",  "width": 110},
        {"fieldname": "return_visitors",  "label": _("Return Visitors"),    "fieldtype": "Int",  "width": 120},
        {"fieldname": "in_followup",      "label": _("In Follow-Up"),       "fieldtype": "Int",  "width": 110},
        {"fieldname": "converted",        "label": _("Converted"),          "fieldtype": "Int",  "width": 100},
        {"fieldname": "lost_contact",     "label": _("Lost Contact"),       "fieldtype": "Int",  "width": 110},
        {"fieldname": "conversion_rate",  "label": _("Conversion Rate %"),  "fieldtype": "Percent","width": 130},
        {"fieldname": "branch",           "label": _("Branch"),             "fieldtype": "Data", "width": 130},
    ]

def get_data(filters):
    branch_cond = "AND branch = %(branch)s" if filters.get("branch") else ""
    from_date   = filters.get("from_date", add_to_date(nowdate(), months=-12))
    to_date     = filters.get("to_date", nowdate())
    period      = filters.get("period", "Monthly")

    if period == "Monthly":
        group_expr = "DATE_FORMAT(date_of_visit, '%%Y-%%m')"
    elif period == "Quarterly":
        group_expr = "CONCAT(YEAR(date_of_visit), '-Q', QUARTER(date_of_visit))"
    else:
        group_expr = "YEAR(date_of_visit)"

    rows = frappe.db.sql(f"""
        SELECT
            {group_expr} AS period,
            COUNT(name)  AS total_visitors,
            SUM(CASE WHEN visit_type = 'First Time Visitor' THEN 1 ELSE 0 END)  AS first_timers,
            SUM(CASE WHEN visit_type = 'Return Visitor'     THEN 1 ELSE 0 END)  AS return_visitors,
            SUM(CASE WHEN conversion_status = 'In Follow-up' THEN 1 ELSE 0 END) AS in_followup,
            SUM(CASE WHEN conversion_status = 'Converted to Member' THEN 1 ELSE 0 END) AS converted,
            SUM(CASE WHEN conversion_status = 'Lost Contact' THEN 1 ELSE 0 END) AS lost_contact,
            branch
        FROM `tabVisitor`
        WHERE date_of_visit BETWEEN %(from_date)s AND %(to_date)s
          {branch_cond}
        GROUP BY {group_expr}, branch
        ORDER BY period
    """, {"from_date": from_date, "to_date": to_date,
          "branch": filters.get("branch","")}, as_dict=True)

    for r in rows:
        r.conversion_rate = round(r.converted / r.total_visitors * 100, 1) if r.total_visitors else 0
    return rows

def get_summary(data):
    total     = sum(d.total_visitors for d in data)
    converted = sum(d.converted for d in data)
    rate      = round(converted / total * 100, 1) if total else 0
    lost      = sum(d.lost_contact for d in data)
    return [
        {"value": total,     "label": _("Total Visitors"),      "datatype": "Int",   "indicator": "blue"},
        {"value": converted, "label": _("Converted to Member"), "datatype": "Int",   "indicator": "green"},
        {"value": rate,      "label": _("Overall Conversion %"),"datatype": "Float", "indicator": "green"},
        {"value": lost,      "label": _("Lost Contact"),        "datatype": "Int",   "indicator": "red"},
    ]

def get_chart(data):
    if not data:
        return {}
    labels    = [d.period for d in data]
    visitors  = [d.total_visitors for d in data]
    converted = [d.converted for d in data]
    return {
        "data": {
            "labels": labels,
            "datasets": [
                {"name": _("Visitors"),  "values": visitors,  "chartType": "bar"},
                {"name": _("Converted"), "values": converted, "chartType": "line"},
            ]
        },
        "type": "axis-mixed",
        "title": _("Visitor Conversion Trend"),
        "colors": ["#3498db","#27ae60"],
        "axisOptions": {"xIsSeries": 1}
    }
